### PR TITLE
Update ubuntu; move to python3; other updates

### PR DESCRIPTION
**Pull request by ghamerly**
This PR does a number of updates:
* problemtools scripts now depend only on python3 rather than python2 (e.g. verifyproblem, problem2html, problem2pdf) -- it removes support for running under python2 (NB: it does not remove python2 as a language for submissions)
* update docker images to run under Ubuntu 20.04 (from 18.04)
* uses pypy3 as the Python 3 interpreter in docker images
* replace libmozjs with nodejs in the language configuration

I have built docker images from this branch which are available here for testing:
* https://hub.docker.com/r/hamerly/problemtools-full
* https://hub.docker.com/r/hamerly/problemtools-icpc
* https://hub.docker.com/r/hamerly/problemtools-minimal
* https://hub.docker.com/r/hamerly/problemtools-build

I would appreciate testing and feedback if anyone has time.

### CODE REVIEW

**Comment by ghamerly** [comment]
I believe these changes are a superset of #168, so if we merge this, we should close that one.

File: problemtools/verifyproblem.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#! /usr/bin/env python2
+#! /usr/bin/env python3
 # -*- coding: utf-8 -*-
 from __future__ import print_function
```
Comment (thorehusfeldt) [inline problemtools/verifyproblem.py:3]: The `__future__` import would be made redundant by this (here and elsewhere).

File: Dockerfile [hunk 2]
```diff
@@ -26,4 +26,4 @@ RUN apt-get update && \
             tidy \
             vim
 
-RUN pip install git+https://github.com/kattis/problemtools
+RUN pip3 install git+https://github.com/ghamerly/problemtools@move-to-py3
```
Comment (mpsijm) [inline Dockerfile:29]: This line should probably be reverted before merging

File: examples/different/submissions/accepted/different.py [hunk 3]
```diff
@@ -1,3 +1,5 @@
+#!/usr/bin/env python2
```
Comment (mpsijm) [inline examples/different/submissions/accepted/different.py:1]: This file could be renamed to `different_py2.py` to avoid confusion

**Review by mpsijm — COMMENTED** [review]
This PR will most likely close #165 :smile:

File: Dockerfile (hunk 2 above)
Comment (ghamerly) [reply Dockerfile]: Fixed, thanks.

File: problemtools/verifyproblem.py (hunk 1 above)
Comment (ghamerly) [reply problemtools/verifyproblem.py]: Fixed, thanks.

**Comment by ghamerly** [comment]
I fixed the issues @mpsijm and @thorehusfeldt identified. I also made a few updates to the imports in the key scripts, and replaced optparse with argparse. I think this is ready for merging.

**Comment by godmar** [comment]
I'm currently working on making a container that combines problem tools + tcardonne's github runner for github CI for naq20. If this PR's landing is imminent, I think I'd want to use it.

**Comment by ghamerly** [comment]
> If this PR's landing is imminent, I think I'd want to use it.

Yes, I think we're going to merge it in the next day or so.

File: admin/docker/Dockerfile.icpc [hunk 4]
```diff
@@ -28,15 +28,20 @@ RUN ln -s /usr/local/kotlinc/bin/* bin/
 
 # Reconfigure problemtools:
 # - Use PyPy for Python 2
+# - Use PyPy for Python 3
 # - Use /usr/local/bin rather than /usr/bin for Kotlin
 RUN mkdir -p /etc/kattis/problemtools
 RUN echo " \n\
 python2: \n\
```
Comment (RagnarGrootKoerkamp) [inline admin/docker/Dockerfile.icpc:35]: Python2 is no longer supported at the worldfinals:
https://icpc.global/worldfinals/programming-environment

File: admin/docker/Dockerfile.icpc [hunk 5]
```diff
@@ -28,15 +28,20 @@ RUN ln -s /usr/local/kotlinc/bin/* bin/
...
     name: 'Python 2 w/PyPy'\n\
     run: '/usr/bin/pypy \"{mainfile}\"'\n\
  \n\
+python3: \n\
+    name: 'Python 3 w/PyPy'\n\
+    run: '/usr/bin/pypy3 \"{mainfile}\"'\n\
+ \n\
```
Comment (RagnarGrootKoerkamp) [inline admin/docker/Dockerfile.icpc:42]: Not sure how overriding works, but the shebang wouldn't be necessary anymore without python 2. (I'd love to just default to python3 by default without shebang, but in general this is bad for backwards compatibility.)

**Review by RagnarGrootKoerkamp — COMMENTED** [review]
Using a different python interpreter for submissions in the ICPC image than in default settings is a bit weird, but making pypy3 the default is a bit annoying as well since not everybody has that installed by default.

File: admin/docker/Dockerfile.icpc (hunk 4 above)
Comment (ghamerly) [reply admin/docker/Dockerfile.icpc:35]: True, but it is possible that judges may still want to use it as a solution language.

**Comment by ghamerly** [comment]
> Using a different python interpreter for submissions in the ICPC image than in default settings is a bit weird, but making pypy3 the default is a bit annoying as well since not everybody has that installed by default.

Sure. But pypy3 *is* in the docker image that this Dockerfile is building.

**Pull request merged**